Sources/Network/NetworkAuthenticator.swift

```diff
@@ -0,0 +1,29 @@
+//
+//  Authenticator.swift
+//  Alicerce
+//
+//  Created by Luís Portela on 03/07/2017.
+//  Copyright © 2017 Mindera. All rights reserved.
+//
+
+import Foundation
+
+public protocol NetworkAuthenticator {
+    func authenticate(request: URLRequest, _ completion: @escaping (URLRequest) -> Void)
```

Comment:
1. I see your point. 
So the "happy path" of `authenticate` is supposed to simply perform the request, correct? 
The "sad path" is when the request fails and then the authentication is made automatically and the request is retried, yes? 
In conclusion, the `completion` should simply signal when either of these cases end. So I'm not understanding how we are sending request creation/validation logic below inside it. (`URLSessionNetworkStack.swift:115`)

2. When I mean requests I mean that the authenticator might need to make more than the one request we give it. What other time consuming operations will it make?